src/lsm/scan_fuzz.zig

```diff
@@ -0,0 +1,1170 @@
+        for (query_parts) |query_part| {
+            switch (query_part) {
+                .field => |field| {
+                    var template = Thing.zeroed();
+                    template.set_index(field.index, field.value);
+
+                    var templates: Templates = .{};
+                    templates.append_assume_capacity(template);
+                    stack.append_assume_capacity(templates);
+                },
+                .merge => |merge| {
+                    var templates: Templates = .{};
+                    const templates_to_merge =
+                        stack.slice()[stack.count() - merge.operand_count ..];
+
+                    switch (merge.operator) {
+                        .union_set => {
+                            // Templates should be added on unions.
+                            for (templates_to_merge) |operand_templates| {
+                                for (operand_templates.const_slice()) |template| {
+                                    templates.append_assume_capacity(template);
+                                }
+                            }
+                        },
+                        .intersection_set => {
+                            // Templates should be merged on intersections,
+                            // see the documentation above for more details.
+                            var matrix_indexes = [_]u32{0} ** query_part_max;
+                            while (matrix_indexes[0] < templates_to_merge[0].count()) {
+                                var template = Thing.zeroed();
+                                for (templates_to_merge, 0..) |template_to_merge, i| {
+                                    template.merge(template_to_merge.get(matrix_indexes[i]));
+                                }
+                                templates.append_assume_capacity(template);
+
+                                for (0..merge.operand_count) |i| {
+                                    const reverse_index = merge.operand_count - 1 - i;
+                                    if (reverse_index == 0 or
+                                        matrix_indexes[reverse_index] <
+                                        templates_to_merge[reverse_index].count() - 1)
+                                    {
+                                        matrix_indexes[reverse_index] += 1;
+                                    } else {
+                                        matrix_indexes[reverse_index] = 0;
+                                        matrix_indexes[reverse_index - 1] += 1;
+                                    }
+                                }
+                            }
+                        },
+                    }
+
+                    stack.truncate(stack.count() - merge.operand_count);
+                    stack.append_assume_capacity(templates);
+                },
+            }
+        }
+
+        assert(stack.count() == 1);
+        return stack.get(0);
+    }
+};
```

Comment:
Hm, statically pre-generating all templates feels very complicated! Can we get by without templates? I think we could generate a random value from the query diretly perhaps? Somethign like this:

```diff
diff --git a/src/lsm/scan_fuzz.zig b/src/lsm/scan_fuzz.zig
index e8bc5929f..ffe3870f9 100644
--- a/src/lsm/scan_fuzz.zig
+++ b/src/lsm/scan_fuzz.zig
@@ -107,6 +107,12 @@ const Thing = extern struct {
         }
     }
 
+    fn merge_all(things: []const Thing) Thing {
+        var result = Thing.zeroed();
+        for (things) |thing| result.merge(thing);
+        return result;
+    }
+
     /// Creates a struct from a template, setting all zeroed fields with random values and
     /// calculating the checksum of the resulting struct.
     fn from_template(template: Thing, random: std.rand.Random, id: u128, timestamp: u64) Thing {
@@ -207,34 +213,12 @@ const Scan = ThingsGroove.ScanBuilder.Scan;
 /// If `lsm_scans_max == x`, then we can have at most x fields and x - 1 merge operations.
 const query_part_max = (constants.lsm_scans_max * 2) - 1;
 
-/// The max number of combinations all indexes can produce.
-/// Combinations Formula: `C(n,r) = n! / (r! * (n - r)!)`,
-/// where `n == lsm_scans_max` and `r == 2` (either assigned or zeroed).
-const template_max: u32 = template_max: {
-    assert(constants.lsm_scans_max >= 2);
-
-    const factorial = struct {
-        fn factorial(comptime n: comptime_int) comptime_int {
-            var result = 1;
-            for (1..n + 1) |i| {
-                result *= i;
-            }
-            return result;
-        }
-    }.factorial;
-
-    break :template_max factorial(constants.lsm_scans_max) /
-        (2 * factorial(constants.lsm_scans_max - 2));
-};
-
 /// The max number of query specs generated per run.
 /// Always generate more than one query spec, since multiple queries can
 /// test both the positive space (results must match the query) and the
 /// negative space (results from other queries must not match).
 const query_spec_max = 8;
 
-const Templates = stdx.BoundedArray(Thing, template_max);
-
 const QueryPart = union(enum) {
     const Field = struct { index: Index, value: u64 };
     const Merge = struct { operator: QueryOperator, operand_count: u8 };
@@ -269,9 +253,6 @@ const QuerySpec = struct {
     query: Query,
     // Ascending or descending.
     reversed: bool,
-    // Matching templates to populate the database.
-    // Zeroed fields are not part of the condition and can be filled with random values.
-    templates: Templates,
     // Number of expected results.
     expected_results: u32,
 
@@ -388,12 +369,10 @@ const QuerySpecFuzzer = struct {
     ) QuerySpec {
         const field_max = self.random.intRangeAtMostBiased(u32, 1, constants.lsm_scans_max);
         const query = self.generate_query(field_max);
-        const templates = get_templates(query.const_slice());
 
         return QuerySpec{
             .prefix = self.prefix,
             .query = query,
-            .templates = templates,
             .reversed = self.random.boolean(),
             .expected_results = 0,
         };
@@ -558,87 +537,6 @@ const QuerySpecFuzzer = struct {
             return index;
         }
     }
-
-    /// Templates are objects that match the condition, such as:
-    /// - The condition `a=1` (from now on represented only as `a` for brevity)
-    ///   generates a template with the field `a` set to `1`, so when inserting
-    ///   objects based on this template, they're expected to match.
-    ///
-    /// - The condition `(a OR b)` generates two templates, one with the
-    ///   field `a` and another one with the field `b` set, so both of them
-    ///   can satisfy the OR clause.
-    ///
-    /// - The condition `(a AND b)` generates only a single template that
-    ///   matches the criteria with both fields `a` and `b` set to the
-    ///   corresponding value.
-    ///
-    /// More complex queries like `(a OR b) AND (c OR d)` need to combine the
-    /// templates generated by each individual condition in such a way any of
-    /// them can satisfy the OR and AND clauses: {a,c},{a,d},{b,c}, and {b,d}.
-    fn get_templates(
-        query_parts: []const QueryPart,
-    ) Templates {
-        var stack = stdx.BoundedArray(Templates, query_part_max){};
-        for (query_parts) |query_part| {
-            switch (query_part) {
-                .field => |field| {
-                    var template = Thing.zeroed();
-                    template.set_index(field.index, field.value);
-
-                    var templates: Templates = .{};
-                    templates.append_assume_capacity(template);
-                    stack.append_assume_capacity(templates);
-                },
-                .merge => |merge| {
-                    var templates: Templates = .{};
-                    const templates_to_merge =
-                        stack.slice()[stack.count() - merge.operand_count ..];
-
-                    switch (merge.operator) {
-                        .union_set => {
-                            // Templates should be added on unions.
-                            for (templates_to_merge) |operand_templates| {
-                                for (operand_templates.const_slice()) |template| {
-                                    templates.append_assume_capacity(template);
-                                }
-                            }
-                        },
-                        .intersection_set => {
-                            // Templates should be merged on intersections,
-                            // see the documentation above for more details.
-                            var matrix_indexes = [_]u32{0} ** query_part_max;
-                            while (matrix_indexes[0] < templates_to_merge[0].count()) {
-                                var template = Thing.zeroed();
-                                for (templates_to_merge, 0..) |template_to_merge, i| {
-                                    template.merge(template_to_merge.get(matrix_indexes[i]));
-                                }
-                                templates.append_assume_capacity(template);
-
-                                for (0..merge.operand_count) |i| {
-                                    const reverse_index = merge.operand_count - 1 - i;
-                                    if (reverse_index == 0 or
-                                        matrix_indexes[reverse_index] <
-                                        templates_to_merge[reverse_index].count() - 1)
-                                    {
-                                        matrix_indexes[reverse_index] += 1;
-                                    } else {
-                                        matrix_indexes[reverse_index] = 0;
-                                        matrix_indexes[reverse_index - 1] += 1;
-                                    }
-                                }
-                            }
-                        },
-                    }
-
-                    stack.truncate(stack.count() - merge.operand_count);
-                    stack.append_assume_capacity(templates);
-                },
-            }
-        }
-
-        assert(stack.count() == 1);
-        return stack.get(0);
-    }
 };
 
 const Environment = struct {
@@ -807,8 +705,6 @@ const Environment = struct {
 
     // TODO: sometimes update and delete things.
     fn populate_things(env: *Environment, query_spec: *QuerySpec) !void {
-        assert(query_spec.templates.count() > 0);
-
         for (0..batch_max) |_| {
             // Total number of objects inserted.
             env.object_count += 1;
@@ -826,28 +722,49 @@ const Environment = struct {
                 continue;
             }
 
-            const template = query_spec.templates.slice()[
-                env.random.intRangeAtMostBiased(
-                    usize,
-                    0,
-                    query_spec.templates.count() - 1,
-                )
-            ];
-
+            const template = env.random_thing_matching_query(query_spec);
             query_spec.expected_results += 1; // Expected objects that match the spec.
-            env.forest.grooves.things.insert(&template.from_template(
+            const thing = template.from_template(
                 env.random,
                 prefix_combine(
                     query_spec.prefix,
                     query_spec.expected_results,
                 ),
                 env.object_count,
-            ));
+            );
+            env.forest.grooves.things.insert(&thing);
         }
 
         try env.commit();
     }
 
+    fn random_thing_matching_query(env: *Environment, query_spec: *QuerySpec) Thing {
+        var stack: [query_part_max]Thing = undefined;
+        var stack_top: usize = 0;
+        for (query_spec.query.slice()) |query_part| {
+            switch (query_part) {
+                .field => |field| {
+                    var thing = Thing.zeroed();
+                    thing.set_index(field.index, field.value);
+                    stack[stack_top] = thing;
+                    stack_top += 1;
+                },
+                .merge => |merge| {
+                    const operands = stack[stack_top - merge.operand_count .. stack_top];
+                    const result = switch (merge.operator) {
+                        .union_set => operands[env.random.uintLessThan(usize, operands.len)],
+                        .intersection_set => Thing.merge_all(operands),
+                    };
+                    stack_top -= merge.operand_count;
+                    stack[stack_top] = result;
+                    stack_top += 1;
+                },
+            }
+        }
+        assert(stack_top == 1);
+        return stack[0];
+    }
+
     fn run_query(env: *Environment, query_spec: *QuerySpec) !void {
         assert(query_spec.expected_results > 0);
 
```